### ISSUE TITLE
Investigate HTTP Status Code 0

### ISSUE THREAD

**Issue by BergWerkGIS** [mapbox/mapbox-unity-sdk#332]
## my current hunch

Based on Unity being single threaded and flaws with our current implementation (yet to be pinpointed):
* ~~too many requests are fired (2 for each tile)~~ not an issue, see below https://github.com/mapbox/mapbox-unity-sdk/issues/332#issuecomment-340758741
* lot of requests block each other leading to timeouts

## evidence

Tried with `Unity@2017.3.0b5` and `Unity@2017.3.0b6`.

Steps to reproduce:
* branch `ErrorCallback`
* `Globe` example
* zoom 6 (remove zoom level safeguard in `GlobeTileProvider.cs`
* add `HandleTileErrorExample.cs` to `Map` game object
* **clear cache**
* `Play`

Results:
* Unity seems to be completely frozen and unresponsive for **~35 seconds** (on my machine)
* Not even a single tile makes it onto the globe:
![image](https://user-images.githubusercontent.com/4549888/32213225-b0288666-be1a-11e7-8eaf-dbcf2589471f.png)
* Nor do they make it into the cache:
![image](https://user-images.githubusercontent.com/4549888/32215514-6f5c83dc-be22-11e7-8987-c9ed5c6b2515.png)
* But a whole bunch (several MB) gets downloaded:
![image](https://user-images.githubusercontent.com/4549888/32215376-18f6e618-be22-11e7-8afa-c033a6bb526a.png)
* All tiles seem to get requested twice:
![image](https://user-images.githubusercontent.com/4549888/32213967-a3446d36-be1d-11e7-87bf-0f930116cba3.png)
* Only after several hundreds of timeouts tiles also get `Status Code 0`:
![image](https://user-images.githubusercontent.com/4549888/32214913-a3cb0dca-be20-11e7-85e5-6a0e052b10b9.png)

**Comment by BergWerkGIS** [comment]
## further observations

* Disabling cache doesn't change anything (for bad or worse). I think it's save to say caching is not part of the problem
* Increasing the timeout parameter to eg 1000 seconds yields **a lot** less failed tiles but also another error (`Cannot connect to destination host`).
![image](https://user-images.githubusercontent.com/4549888/32217247-ae4eafe8-be27-11e7-905d-e2d5605b6fad.png)
* With increased timeout and several runs (till all tiles are in the cache) the globe finally displays as expected:
![image](https://user-images.githubusercontent.com/4549888/32217279-c9e5118e-be27-11e7-963a-8f9b2f56e955.png)
but memory usage is very high:
![image](https://user-images.githubusercontent.com/4549888/32217258-b71af1a4-be27-11e7-82b3-dc733bf07ee8.png)
* exporting to `Windows Standalone x64` player yields comparable results:
  * application seems to frozen
  * with default timeout (10s) no tiles are displayed
  * with timeout 1000s and empty cache memory usage also climbs to ~8GB but then the player _sometimes_ crashes with:
    ```
    Read from location 00000110 caused an access violation.
    ```
  * after several runs (till all tiles are in the cache) globe looks like expected. Again with ~8GB memory usage.
* using `Experimental .NET 4.6 Scripting Runtime` yields comparable results

**Comment by BergWerkGIS** [comment]
Analyzing web requests from outside of Unity (with Fiddler) shows that tiles **don't** get requested twice, seems it's only their error that shows up twice in the console.

Example session profiling web requests :

* at zoom 5
* should be 1,024 tiles
* Fiddler shows that 938 requests were sent and received
* leaves a difference of 86 tiles (1,024-938) that failed
* console shows 172 error messages which is double the number of failed tiles (86*2)

**Comment by BergWerkGIS** [comment]
## 1 exception vs 2 exceptions per tile:

**1 exception**

These seem to be in an intermediate state and ran into the timeout when a response was already received but not processed.
Thus response does have a valid status code.
All of those that I've inspected had `200/OK` **and** had at least some data attached.
Some of that data was the full tile and some not.

**2 exceptions (with additional Status Code `0`)**

These seem to be the ones that ran into the timeout before even being fired or shortly thereafter.
Thus `Status Code 0`: no response from server included.

## `UnityWebRequest.timeout` / `UnityWebRequest.SendWebRequest()`

I've also switched to the new `UnityWebRequest.timeout` and `UnityWebRequest.SendWebRequest()` but that didn't change anything.
https://github.com/mapbox/mapbox-unity-sdk/commit/fb994f0abdccdbc27e7ff1a0a6d3498bbd6ee0d5

**Comment by david-rhodes** [comment]
Possible duplicate exceptions as a results of #335?

**Comment by BergWerkGIS** [comment]
Problem occurs when too many tile requests are fired at once.

Going to solve this via an queue wrapper around tile requests that throttles request according to the rate limits returned by API headers.

**Comment by jekelija** [comment]
Did this actually solve your problem? We are running into a similar issue in our code, but when i try to reproduce it using a test set that just fires off 200 requests at once, we do not get the error, so we assumed that it didnt have to do with many requests firing at once?
